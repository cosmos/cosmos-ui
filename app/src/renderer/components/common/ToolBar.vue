<template>
  <div class="tool-bar">
    <a
      v-tooltip.bottom="'Back'"
      :disabled="user.history.length === 0"
      class="back"
      @click="back"
    >
      <i class="material-icons">arrow_back</i>
    </a>
    <a
      v-if="!!refresh"
      v-tooltip.bottom="'Refresh'"
      :disabled="!refresh.connected"
      class="refresh-button"
      @click="refresh.connected && refresh.refresh()"
    >
      <i class="material-icons">refresh</i>
    </a>
    <a
      v-if="!!searching"
      v-tooltip.bottom="'Search'"
      :disabled="!searching.somethingToSearch"
      class="search-button"
      @click="searching.setSearch()"
    >
      <i class="material-icons">search</i>
    </a>
    <slot />
    <a v-tooltip.bottom="'Help'" class="help" @click="enableModalHelp">
      <i class="material-icons">help_outline</i>
    </a>
    <router-link
      v-if="config.devMode"
      id="settings"
      v-tooltip.bottom="'Preferences'"
      to="/preferences"
    >
      <i class="material-icons">settings</i>
    </router-link>
    <a id="signOut-btn" v-tooltip.bottom.end="'Sign Out'" @click="signOut">
      <i class="material-icons">exit_to_app</i>
    </a>
  </div>
</template>

<script>
import { mapGetters, mapMutations } from "vuex"
export default {
  name: `tool-bar`,
  props: {
    refresh: {
      type: Object,
      default: undefined
    },
    searching: {
      type: Object,
      default: undefined
    }
  },
  computed: {
    ...mapGetters([`user`, `lastPage`, `config`]),
    searchEnabled() {
      return !!this.searching
    },
    somethingToSearch() {
      return this.searching && this.searching.somethingToSearch()
    },
    setSearch() {
      return this.searching && this.searching.setSearch()
    }
  },
  methods: {
    ...mapMutations([`pauseHistory`, `popHistory`]),
    back() {
      if (!this.lastPage) return
      this.pauseHistory(true)
      this.$router.push(this.lastPage, () => {
        this.popHistory()
        this.pauseHistory(false)
      })
    },
    enableModalHelp() {
      this.$store.commit(`setModalHelp`, true)
    },
    signOut() {
      this.$store.dispatch(`signOut`)
    }
  }
}
</script>
<style>
.tm-tool-bar {
  align-self: start;
}
.tm-page-header-text {
  padding-right: 1rem;
}

.tool-bar a {
  padding-left: 0.5rem;
  position: relative;
  top: 1rem;
  right: 1rem;
  color: var(--dim);
}

.tool-bar a:hover {
  cursor: pointer;
  color: var(--bright);
}
</style>
