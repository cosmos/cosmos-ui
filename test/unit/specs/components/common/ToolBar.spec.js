import setup from "../../../helpers/vuex-setup"
import ToolBar from "common/ToolBar"
describe(`ToolBar`, () => {
  let wrapper, store
  const { mount } = setup()

  beforeEach(() => {
    const instance = mount(ToolBar)
    wrapper = instance.wrapper
    store = instance.store
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`sets the helper modal`, () => {
    wrapper.vm.enableModalHelp()
    expect(store.state.session.modals.help.active).toBe(true)
  })

  it(`call dispatch to sign the user out`, () => {
    const $store = { dispatch: jest.fn() }
    const self = { $store, $router: { push: jest.fn() } }
    ToolBar.methods.signOut.call(self)
    expect($store.dispatch).toHaveBeenCalledWith(`signOut`)
  })

  it(`opens session modal`, () => {
    const $store = { commit: jest.fn() }
    const self = { $store }
    ToolBar.methods.signIn.call(self)
    expect($store.commit).toHaveBeenCalledWith(`setSessionModalView`, `welcome`)
    expect($store.commit).toHaveBeenCalledWith(`toggleSessionModal`, true)
  })

  it(`check if search should be Enabled`, () => {
    expect(ToolBar.computed.searchEnabled.call({ searching: true })).toBe(true)
    expect(ToolBar.computed.searchEnabled.call({})).toBe(false)
  })

  it(`fails to check if there is somethingToSearch`, () => {
    expect(ToolBar.computed.somethingToSearch.call({})).toBe(undefined)
  })
  it(`succeed in checking if there is somethingToSearch`, () => {
    const somethingToSearch = jest.fn()
    const localThis = { searching: { somethingToSearch } }
    ToolBar.computed.somethingToSearch.call(localThis)
    expect(somethingToSearch).toHaveBeenCalled()
  })

  it(`fails to setSearch`, () => {
    expect(ToolBar.computed.setSearch.call({})).toBe(undefined)
  })

  it(`succeed in setSearch`, () => {
    const setSearch = jest.fn()
    const localThis = { searching: { setSearch } }
    ToolBar.computed.setSearch.call(localThis)
    expect(setSearch).toHaveBeenCalled()
  })
})
