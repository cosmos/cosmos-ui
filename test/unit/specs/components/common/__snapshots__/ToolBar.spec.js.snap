// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolBar has the expected html structure 1`] = `
<div
  class="tool-bar"
>
  <!---->
    
  <a
    id="signOut-btn"
  >
    <i
      class="material-icons"
    >
      
      exit_to_app
    
    </i>
  </a>
   
  <!---->
</div>
`;
